### PR TITLE
implement drag-and-drop kanban board

### PR DESCRIPTION
**Pull request by lucaslyl**
This task is to develop a **DndKanbanBoard** component that implements a drag-and-drop Kanban board using ember-draggable-modifiers. 

- allowing users to create their own functionality for column headers and design custom draggable cards. This flexibility enables developers to implement unique actions or menus within column headers.
- Additionally, users should have the freedom to define their own card designs with custom style.

https://github.com/user-attachments/assets/e9f33962-d0e8-4a43-9993-def752eee3d7

### CODE REVIEW

**Review by habdelra — APPROVED** [review]
nice! just need to solve the fastboot issue

**Comment by habdelra** [comment]
Looks like the test failures are legit. If you look at the server log that was uploaded from the test failures you'll see that there appears to be some browser specific code that is now running in nodejs. Specifically:

```
Realm server listening on port 4201

worker: Error raised during indexing has likely stopped the indexer ReferenceError: navigator is not defined
    at isSafari (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:6252:20)
    at wrapped (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:5753:23)
    at fixSafari (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:5862:60)
    at Module.../../../../../node_modules/.pnpm/@atlaskit+pragmatic-drag-and-drop@1.3.1/node_modules/@atlaskit/pragmatic-drag-and-drop/dist/esm/util/changing-window/count-events-for-safari.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:5941:3)
    at __webpack_require__ (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.e294b0957394a7e22a16.js:23:42)
    at Module.../../../../../node_modules/.pnpm/@atlaskit+pragmatic-drag-and-drop@1.3.1/node_modules/@atlaskit/pragmatic-drag-and-drop/dist/esm/util/changing-window/is-leaving-window.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:6058:82)
    at __webpack_require__ (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.e294b0957394a7e22a16.js:23:42)
    at Module.../../../../../node_modules/.pnpm/@atlaskit+pragmatic-drag-and-drop@1.3.1/node_modules/@atlaskit/pragmatic-drag-and-drop/dist/esm/ledger/lifecycle-manager.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:4622:97)
    at __webpack_require__ (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.e294b0957394a7e22a16.js:23:42)
    at Module.../../../../../node_modules/.pnpm/@atlaskit+pragmatic-drag-and-drop@1.3.1/node_modules/@atlaskit/pragmatic-drag-and-drop/dist/esm/make-adapter/make-adapter.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:5022:83)
    at __webpack_require__ (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.e294b0957394a7e22a16.js:23:42)
    at Module.../../../../../node_modules/.pnpm/@atlaskit+pragmatic-drag-and-drop@1.3.1/node_modules/@atlaskit/pragmatic-drag-and-drop/dist/esm/adapter/element-adapter.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:3414:84)
    at __webpack_require__ (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.e294b0957394a7e22a16.js:23:42)
    at Module.../../../../../node_modules/.pnpm/@atlaskit+pragmatic-drag-and-drop@1.3.1/node_modules/@atlaskit/pragmatic-drag-and-drop/dist/esm/entry-point/element/adapter.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:4011:82)
    at __webpack_require__ (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.e294b0957394a7e22a16.js:23:42)
    at Module.../../../../../node_modules/.pnpm/ember-draggable-modifiers@1.0.0_@babel+core@7.24.3_ember-source@5.4.1/node_modules/ember-draggable-modifiers/dist/modifiers/draggable-item.js (/tmp/tmp-3745-MUv3E5mw4c2K/assets/chunk.eef8c6cd3847f29b44f1.js:89124:107)
```
It looks like the drag and drop library we are using is hostile to fastboot (this happens during indexing time when we pre-render the HTML). Perhaps we should add a fastboot guard in our modifiers so that we don't call out of this library if we are in fastboot. There is a chance, though, that this library might be eagerly performing these browser specific checks as a result of simply importing it (e.g. "side effects"). In that case I think we'd need to dynamically import this lib if you still see these errors after statically checking to see if you are running in fastboot.

This is what a fastboot guard looks like: https://github.com/cardstack/boxel/blob/main/packages/host/app/components/card-prerender.gts#L31-L36

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+    this.clearHoveredState();
+  }
+
+  <template>
+    <div class='draggable-container' {{on 'dragend' this.clearHoveredState}}>
+      {{#each this.columns as |column columnIndex|}}
+        <div
+          class='column'
+          {{DndDropTargetModifier
```
Comment (habdelra) [inline packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: if we are in fastboot we will not want to include this modifier

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+            {{#if (this.isDropZoneTargeted columnIndex)}}
+              <div class='column-drop-zone-overlay'>
+                Board ordered by Priority
+              </div>
+            {{/if}}
+
+            {{#each column.cards as |card|}}
+              <div
+                class='draggable-card'
+                {{DndSortableItemModifier
```
Comment (habdelra) [inline packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: if we are in fastboot we'll not want to include this modifier

**Comment by habdelra** [comment]
Note that I also filed an issue against this addon to add fastboot support (but that should not preclude us from addressing the issue on our own) https://github.com/alvarocastro/ember-draggable-modifiers/issues/1

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts (hunk 1 above)
Comment (lucaslyl) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: Thanks for pointing that out! I’ve already implemented dynamic import for the module, and it seems fixed the error.

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+      const DndDropTargetModifier = await import(
+        'ember-draggable-modifiers/modifiers/drop-target'
+      );
+
+      // @ts-expect-error Dynamic imports are only supported when the '--module' flag is set to 'es2020', 'es2022', 'esnext', 'commonjs', 'amd', 'system', 'umd', 'node16', or 'nodenext'
+      const DndSortableItemModifier = await import(
+        'ember-draggable-modifiers/modifiers/sortable-item'
+      );
+      // @ts-expect-error Dynamic imports are only supported when the '--module' flag is set to 'es2020', 'es2022', 'esnext', 'commonjs', 'amd', 'system', 'umd', 'node16', or 'nodenext'
+      const arrayUtils = await import('ember-draggable-modifiers/utils/array');
```
Comment (habdelra) [inline packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: i believe we are using `es2022` everywhere else. is there a reason not to here (@lukemelia @ef4 )?

Comment (habdelra) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts:86]: probably worth adding a comment that we are lazily loading these modifiers so they play nice in fastboot

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+  constructor(title: string, cards: Card[] = []) {
+    this.title = title;
+    this.cards = cards;
+  }
+}
+
+export default class DndKanbanBoard extends Component<
+  DndKanbanBoardSignature<Column>
+> {
+  @tracked isModifiersLoaded = false;
```
Comment (habdelra) [inline packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: better english: `areModifiersLoaded`

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+      this.DndDropTargetModifier = DndDropTargetModifier.default;
+      this.DndSortableItemModifier = DndSortableItemModifier.default;
+      this.insertAfter = arrayUtils.insertAfter;
+      this.insertAt = arrayUtils.insertAt;
+      this.insertBefore = arrayUtils.insertBefore;
+      this.removeItem = arrayUtils.removeItem;
+
+      this.isModifiersLoaded = true;
+    } catch (e) {
+      console.error('Error loading modifiers:', e);
```
Comment (habdelra) [inline packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: why are we swallowing this error? shouldn't we let it explode?

Comment (habdelra) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: also keep the try/catch scope as small as possible--only wrap the area that you expect to throw. there is no need to wrap all the assignments on 88-94 in a try/catch

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts (hunk 3 above)
Comment (lukemelia) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: I don't think so, but I'd defer to @ef4.

File: packages/boxel-ui/addon/src/components/drag-and-drop/index.gts (hunk 1 above)
Comment (lukemelia) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: Dynamic imports seems like a weird way to solve this problem. I think we could use [the `modifier` helper](https://api.emberjs.com/ember/5.12/classes/Ember.Templates.helpers/methods/modifier?anchor=modifier) to conditionally apply a modifier when not in Fastboot.

Comment (lucaslyl) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: @lukemelia  I have used the modifier helper in a previous revision, but as mentioned by @FadhlanR, the ember-draggable-modifiers is not compatible with Fastboot. When I statically imported the modifier helpers, it generated the error (Error raised during indexing has likely stopped the indexer: ReferenceError: navigator is not defined). The error only disappeared after we switched to dynamic import. @habdelra , do you have any suggestions other than using dynamic import?

Comment (habdelra) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: yeah, i don't think we have a choice, the underlying lib that ember-drag-and-drop uses has side effects that need the browser context when the modules are evaluated. we have to dynamically load these modules as the very  act of importing them runs the side effects that blow up fastboot. statically importing these deps will evaluate the side effects before we have a chance to do anything (i.e. modifier helper is way too late to deal with this). dynamic importing is the only way to safely handle side effects like this.

Comment (lucaslyl) [reply packages/boxel-ui/addon/src/components/drag-and-drop/index.gts]: @lukemelia Hi luke, we still using dynamic import for now, i will capture a ticket to revisit it.

**Pull request merged**